### ISSUE TITLE
Nvidia H100 nvproxy: unknown control command 0x20801230

### ISSUE THREAD

**Issue by derpsteb** [google/gvisor#9846]
### Description

Hi,
I am trying to use an Nvidia H100 with gVisor's nvproxy but I am hitting the error from the title when running a simple example. The example is in the reproductions steps. It is taken from the [commit](https://github.com/google/gvisor/commit/333bd93e1adf514cbac7a53f5fc0b42e4708803b) that adds support for the h100.
The actual user-facing error that I am seeing with other tests I am running as well (e.g. running queryDevice inside a container) is sth along the lines of `no CUDA-capable device is detected`.

When looking at the strace logs of the failing containers the only error I can spot is the error from the title. Is this expected? Tbh, I am expecting a version mismatch somewhere in my setup to be responsible for the problem. But the error made me a bit suspicious. 

The unknown control command seems to be [this](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/common/sdk/nvidia/inc/ctrl/ctrl2080/ctrl2080gr.h#L1533) one: `NV2080_CTRL_CMD_GR_GET_SM_ISSUE_RATE_MODIFIER`

Thank you for looking into this!

### Steps to reproduce
- This works: `docker run --gpus all nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda11.7.1-ubi8`
- This doesn't work: `docker run --runtime=runsc --gpus all nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda11.7.1-ubi8`

- qemu VM running ubuntu 22.04
- cuda versions tested: `525.60.13`, `535.54.03`, `535.86.10`, `535.104.05`. At least nr 1, 2 and 4 of those should be supported according to `runsc nvproxy list-supported-drivers`

### runsc version

```shell
runsc version release-20231218.0
spec: 1.1.0-rc.1
```

### docker version (if using docker)

```shell
Client: Docker Engine - Community
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.20.10
 Git commit:        afdd53b
 Built:             Thu Oct 26 09:07:41 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:07:41 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.26
  GitCommit:        3dd1e886e55dd695541fdcd67420c2888645a495
 runc:
  Version:          1.1.10
  GitCommit:        v1.1.10-0-g18a0cb0
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### uname

Linux guest 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### runsc debug logs (if available)

This is just the section surrounding the `nvproxy: unknown control command ...` output. Happy to extend with the full log if necessary.
```shell
I1229 09:57:11.030378   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480) = 0 (0x0) (21.883µs)
I1229 09:57:11.030403   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480)
D1229 09:57:11.030410   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.030417   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x2080122b
I1229 09:57:11.030431   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480) = 0 (0x0) (20.731µs)
I1229 09:57:11.030474   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480)
D1229 09:57:11.030481   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.030488   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x2080122b
I1229 09:57:11.030504   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480) = 0 (0x0) (22.494µs)
I1229 09:57:11.030535   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480)
D1229 09:57:11.030543   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.030551   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x2080122b
I1229 09:57:11.030566   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480) = 0 (0x0) (23.855µs)
I1229 09:57:11.030594   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113470)
D1229 09:57:11.030600   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.030607   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x20801227
I1229 09:57:11.030624   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113470) = 0 (0x0) (23.606µs)
I1229 09:57:11.030659   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113420)
D1229 09:57:11.030666   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.030674   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x20802a0a
I1229 09:57:11.032078   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113420) = 0 (0x0) (1.410376ms)
I1229 09:57:11.032131   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480)
D1229 09:57:11.032138   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.032146   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x20801230
W1229 09:57:11.032153   13265 frontend.go:521] [   2:   2] nvproxy: unknown control command 0x20801230 (paramsSize=32)
I1229 09:57:11.032162   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe51113480) = 0 (0x0) errno=22 (invalid argument) (21.543µs)
I1229 09:57:11.032212   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe511145e0)
D1229 09:57:11.032219   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x00000029, argSize = 0x00000010
I1229 09:57:11.033586   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe511145e0) = 0 (0x0) (1.364655ms)
I1229 09:57:11.033642   13265 strace.go:561] [   2:   2] python E close(0x7 /dev/nvidia0)
I1229 09:57:11.033664   13265 strace.go:599] [   2:   2] python X close(0x7 /dev/nvidia0) = 0 (0x0) (13.83µs)
I1229 09:57:11.033695   13265 strace.go:561] [   2:   2] python E close(0x6 /dev/nvidia0)
I1229 09:57:11.033707   13265 strace.go:599] [   2:   2] python X close(0x6 /dev/nvidia0) = 0 (0x0) (4.577µs)
I1229 09:57:11.033719   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe511134b0)
D1229 09:57:11.033726   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x0000002a, argSize = 0x00000020
D1229 09:57:11.033735   13265 frontend.go:488] [   2:   2] nvproxy: control command 0x216
I1229 09:57:11.033753   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc020462a, 0x7ebe511134b0) = 0 (0x0) (26.861µs)
I1229 09:57:11.033792   13265 strace.go:561] [   2:   2] python E close(0x5 /dev/nvidia0)
I1229 09:57:11.033803   13265 strace.go:599] [   2:   2] python X close(0x5 /dev/nvidia0) = 0 (0x0) (4.447µs)
I1229 09:57:11.033848   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe51116470)
D1229 09:57:11.033855   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x00000029, argSize = 0x00000010
I1229 09:57:11.033877   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe51116470) = 0 (0x0) (20.682µs)
I1229 09:57:11.033926   13265 strace.go:567] [   2:   2] python E ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe51116470)
D1229 09:57:11.033937   13265 frontend.go:154] [   2:   2] nvproxy: frontend ioctl: nr = 0x00000029, argSize = 0x00000010
I1229 09:57:11.033968   13265 strace.go:605] [   2:   2] python X ioctl(0x3 /dev/nvidiactl, 0xc0104629, 0x7ebe51116470) = 0 (0x0) (30.066µs)
I1229 09:57:11.033995   13265 strace.go:561] [   2:   2] python E close(0x3 /dev/nvidiactl)
I1229 09:57:11.034009   13265 strace.go:599] [   2:   2] python X close(0x3 /dev/nvidiactl) = 0 (0x0) (7.051µs)
I1229 09:57:11.034025   13265 strace.go:567] [   2:   2] python E ioctl(0x4 /dev/nvidia-uvm, 0x30000002, 0x0)
D1229 09:57:11.034032   13265 uvm.go:135] [   2:   2] nvproxy: uvm ioctl 0x30000002
```

**Comment by ayushr2** [comment]
Thanks for the detailed report.

What Nvidia driver version are you using? I had tested https://github.com/google/gvisor/commit/333bd93e1adf514cbac7a53f5fc0b42e4708803b on version `525.60.13` (as the commit describes).

> When looking at the strace logs of the failing containers the only error I can spot is the error from the title. Is this expected?

This maybe the cause of the failure, but need to confirm. Sometimes I have seen GPU workloads work fine in spite of such warnings.

**Comment by ayushr2** [comment]
Also if you get a chance, could you patch https://github.com/google/gvisor/pull/9847 and see if it works?

H100s are hard to get a hold of right now, so hard to test. But let me know if it works, and what driver versions you are using.

**Comment by derpsteb** [comment]
Heyhey. Thanks a lot for the quick feedback! 
- Testing cuda version `525.60.13` on a 5.15 kernel (as indicated in the compatibility info of that version) works if the confidential computing mode of the GPU is disabled. As mentioned in the original ticket.
- Testing on cuda version `535.104.05` on a 6.2.0 kernel works, if I add your patch and a few more missing commands. CC mode disabled.
- Enabling CC mode on those versions brings me into a infinite loop. Presumably during `cudaMalloc`. I will create a minimal example and open another issue for that.

You can find my patches [here](https://github.com/derpsteb/gvisor/commit/61a27388f985ceba60db77fdc0feb9b4198ffcc5). I will open a PR once CC mode works.

**Comment by ayushr2** [comment]
Awesome, seems like you are on top of this. Happy to see nvproxy contributions coming through! Please tag me in the PR once you open it.

**Comment by ayushr2** [comment]
I have re-sent https://github.com/google/gvisor/pull/9847 for review in the meantime (to unblock Modal). I don't see `NV2080_CTRL_CMD_GR_GET_SM_ISSUE_RATE_MODIFIER` being added in https://github.com/derpsteb/gvisor/commit/118d69a836487a5eed782196b2e6af8d40527392, so not sure if @derpsteb intends to add this in his series.

@luiscape likely you will hit more unimplemented commands/ioctls. I think it is more efficient to run such GPU application natively and collect all ioctls used by application and implement them at once. We need tooling to do this. I intended to discuss this in #9921. Lets move that discussion there.

**Comment by derpsteb** [comment]
Hey @luiscape. I will add a PR next week. We also used ayushr2's patch. So either it's merged independently or I pull it into the PR.

**Comment by luiscape** [comment]
@derpsteb nice. Let us know if you need any help. We are currently at `535.129.03` so happy to test there.
